chart/kubeapps/templates/apprepository/apprepositories-secret.yaml

```diff
@@ -21,9 +21,15 @@ data:
   ca.crt: |-
     {{ .caCert | b64enc }}
   {{- end }}
+  {{- $authorizationHeader := "" }}
   {{- if .authorizationHeader }}
+    {{- $authorizationHeader = .authorizationHeader | b64enc }}
+  {{- else if .basicAuth }}
+    {{- $authorizationHeader = printf "Basic %s" ((printf "%s:%s" .basicAuth.user .basicAuth.password) | b64enc) }}
+  {{- end }}
+  {{- if $authorizationHeader }}
   authorizationHeader: |-
-    {{ .authorizationHeader | b64enc }}
+    {{ $authorizationHeader | b64enc }}
```

Comment:
Isn't `$authorizationHeader` now double-base64 encoded? I'm guessing you want to remove the b64 encoding on lines 26 and 28 (or here and below on line 56).

Reply:
Hello
good catch there were one to many b64enc call, however for the basic authentication there are 2 b64enc involved:
> If the "Basic" authentication scheme is used, the credentials are constructed by first combining the username and the password with a colon (aladdin:opensesame), then by encoding the resulting string in base64 (YWxhZGRpbjpvcGVuc2VzYW1l). 
source: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization

then when building the secret, [the values for all keys in the data field have to be base64-encoded strings](https://kubernetes.io/docs/concepts/configuration/secret/#overview-of-secrets).

Here is an example:
```
$ helm template . --set apprepository.initialRepos[0].name=local --set apprepository.initialRepos[0].url=http://localhost --set apprepository.initialRepos[0].basicAuth.user=myuser --set apprepository.initialRepos[0].basicAuth.password=mypassword
```
yields:
```yaml
# Source: kubeapps/templates/apprepository/apprepositories-secret.yaml
apiVersion: v1
kind: Secret
metadata:
  name: apprepo-local-secrets
  namespace: "default"
  labels:
    app.kubernetes.io/name: kubeapps
    helm.sh/chart: kubeapps-7.4.1-dev0
    app.kubernetes.io/instance: RELEASE-NAME
    app.kubernetes.io/managed-by: Helm
data:
  authorizationHeader: |-
    QmFzaWMgYlhsMWMyVnlPbTE1Y0dGemMzZHZjbVE9
```
which is decoded as follows:
```bash
$ echo -n QmFzaWMgYkdWa1oyVnlhR1ZzYldSbGRqb3libGhHWkdGb01taElaV2RwWm1neFpHNUJSZz09 | base64 -d
Basic bGVkZ2VyaGVsbWRldjoyblhGZGFoMmhIZWdpZmgxZG5BRg==
$ echo -n QmFzaWMgYlhsMWMyVnlPbTE1Y0dGemMzZHZjbVE9 | base64 -d | sed 's/Basic //' | base64 -d
myuser:mypassword
```

whereas
```bash
$ helm template . --set apprepository.initialRepos[0].name=local --set apprepository.initialRepos[0].url=http://localhost --set apprepository.initialRepos[0].authorizationHeader="Bearer mytoken"
```
yields
```
# Source: kubeapps/templates/apprepository/apprepositories-secret.yaml
apiVersion: v1
kind: Secret
metadata:
  name: apprepo-local-secrets
  namespace: "default"
  labels:
    app.kubernetes.io/name: kubeapps
    helm.sh/chart: kubeapps-7.4.1-dev0
    app.kubernetes.io/instance: RELEASE-NAME
    app.kubernetes.io/managed-by: Helm
data:
  authorizationHeader: |-
    QmVhcmVyIG15dG9rZW4=
```
which is decoded as follows:
```bash
$ echo -n QmVhcmVyIG15dG9rZW4= | base64 -d
Bearer mytoken
```

Reply:
I initially thought the same, but I think it's right as is?

The header should be b64-ed itself, and the Basic auth is `b64(user:pass)`, so, the whole thing is: `b64("Basic b64(<user>:<pass>)")` , no?

Reply:
For the basic auth it was, but an extra one had been added to the non-basic auth path. Thanks for the fix in https://github.com/kubeapps/kubeapps/pull/3372/commits/97830badae6369efaa9d9f9e1458c53a62c3489c Herve.